### PR TITLE
Test case for competitor section

### PR DESCRIPTION
**Pull request by ShihaMohan235**
- spacing test case
- card redirection

### CODE REVIEW

File: travel-and-expense-software/test_travel_expense_software.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        "//section[contains(@class, 'fyle-vs-competitors')]//a[contains(@class, 'competitor-cards')]//p[contains(text(), 'Chromeriver')]//parent::a"
+    ]
+    for i, card_xpath in enumerate(card_list):
+        card = browser.find(card_xpath, scroll=True)
+        browser.click_element(card)
+        assert browser.get_current_url() == card_url[i], "redirection url is not correct"
+        browser.back()
+
+@pytest.mark.parametrize('browser', [('mobile_1')], indirect=True)
+def test_competitor_section_card_redirection_mobile(browser):
```
Comment (Krishnakumarskr) [inline travel-and-expense-software/test_travel_expense_software.py:79]: are we assigning the card_url and list twice for desktop and mobile?  if the same variable is required for both mobile and desktop test cases, we have to look for some optimized approach.

Comment (jatzz10) [inline travel-and-expense-software/test_travel_expense_software.py:79]: I think we can use `assert_cards_redirection()` function for this test case. Can you check once?

Comment (ShihaMohan235) [reply travel-and-expense-software/test_travel_expense_software.py:85]: That can be used where on clicking when URL opens in a new tab. In this case, its opening in the same tab

File: travel-and-expense-software/test_travel_expense_software.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        assert_spacing_top(card, '20')
+        assert_spacing_below(card, '20')
+    for subtext in card_sub_text:
+        assert_spacing_top(subtext, '15')
+        assert_spacing_below(subtext, '15')
+        assert_spacing_right(subtext, '0')
+        assert_spacing_left(subtext, '0')
+
+@pytest.mark.parametrize('browser', [('desktop_1')], indirect=True)
+def test_competitor_section_card_redirection(browser):
```
[the remark this answers is not in the document]
Comment (ShihaMohan235) [reply travel-and-expense-software/test_travel_expense_software.py:65]: Didnt move this to common py, as only 1 or 2 pages maximum will be using this section and this code can be reused in those particular pages

[the remark this answers is not in the document]
Comment (ShihaMohan235) [reply travel-and-expense-software/test_travel_expense_software.py:65]: calling a common fn in 89th line

File: travel-and-expense-software/test_travel_expense_software.py (hunk 1 above)
Comment (jatzz10) [reply travel-and-expense-software/test_travel_expense_software.py]: I think we can tweak this function to work for both the cases. I will also take a look at this today.

File: common/asserts.py [hunk 3]
```diff
@@ -93,6 +128,14 @@ def assert_cards_redirection(browser, cards_xpath, redirect_to_urls):
             browser.scroll_up_or_down(300)
         sleep(2)
 
+def assert_cards_redirection_same_tab(browser, cards_list, url_list):
```
Comment (jatzz10) [inline common/asserts.py:131]: Shiha, we can reuse the above function. Can give another param in above function like this:
```
def assert_cards_redirection(browser, cards_xpath, redirect_to_urls, same_tab=False):
    cards = browser.find_many(xpath=cards_xpath)
    assert len(cards) > 0, 'Wrong xpath given for cards'
    for card in cards:
        browser.click_element(card)
        if same_tab:
            assert browser.get_current_url() in redirect_to_urls, 'Redirecting to wrong page'
            browser.back()
        else:
            browser.switch_tab_next(1)
            assert browser.get_current_url() in redirect_to_urls, 'Redirecting to wrong page'
            browser.close_windows()
        if browser.is_desktop() is False:
            browser.scroll_up_or_down(300)
        sleep(2)
```

If we want to call for redirection in same tab, we can call like this:
`assert_cards_redirection(browser, cards_xpath, redirect_to_urls, same_tab=True)`

Comment (jatzz10) [reply common/asserts.py]: Is the above snippet mentioned in the comment not working for you? It seems you are still using some redundant code right now in this function.

File: common/asserts.py [hunk 4]
```diff
@@ -81,17 +116,25 @@ def assert_collapsible_feature_comparison_table(browser):
...
-        assert browser.get_current_url() in redirect_to_urls, 'Redirecting to wrong page'
-        browser.close_windows()
-        if browser.is_desktop() is False:
-            browser.scroll_up_or_down(300)
-        sleep(2)
+def assert_cards_redirection(browser, cards_xpath, redirect_to_urls, same_tab=False):
+    if same_tab:
+        for i, card_elem in enumerate(cards_xpath):
+           card = browser.find(card_elem, scroll=True)
+           browser.scroll_up_or_down(-100)
```
Comment (Krishnakumarskr) [inline common/asserts.py:123]: So up have +ve value and down have -ve value, correct? but it was confusing when looking for the first time.

File: expense_page/test_expense_features.py [hunk 5]
```diff
@@ -1,77 +1,77 @@
...
+# @pytest.mark.parametrize('browser', [('desktop_1'), ('mobile_1')], indirect=True)
+# def test_success(browser):
+#     assert_success(browser)
 
-@pytest.mark.parametrize('browser', [('desktop_1')], indirect=True)
-def test_feature_scroll(browser):
-    browser.click(xpath="//section[@class='feature-hero']//a[text()='Expenses']")
-    e = browser.find(xpath="//section[@id='expense-tracking']")
-    assert abs(e.location['y'] - browser.current_scroll_position()) <= 30, 'Not scrolling to the Expenses feature'
+# @pytest.mark.parametrize('browser', [('desktop_1')], indirect=True)
```
Comment (Krishnakumarskr) [inline expense_page/test_expense_features.py:32]: is there any use of these comments in the file?

**Review by jatzz10 — APPROVED** [review]

**Review by Krishnakumarskr — APPROVED** [review]

**Pull request closed without merging**